### PR TITLE
[mono] Use correct cast_class for IntPtr[]

### LINKED ISSUE #97145: [mono] Incorrect cast_class for IntPtr[]
### Description

mono repo has a  intptr array Unit Test code:
https://github.com/dotnet/runtime/blob/13ab6eb58e1c16856cfd3bff2fe69da13d6642af/src/mono/mono/mini/arrays.cs#L693
This test case argues that an integer type array  should be able to convert the pointer type array
It's produce difference value
.Net Framwork = 0
.Net Core = 5

It like there's something wrong with the .Net Framework

### PR DESCRIPTION
**Pull request by steveisok**
This change fixes the behavior to be like modern .NET instead of .NET Framework.

Fixes https://github.com/dotnet/runtime/issues/97145

### CODE REVIEW

**Review by lambdageek — APPROVED** [review]

**Review by kg — APPROVED** [review]
Looks right, but I'm not 100% certain.

**Comment by lambdageek** [comment]
I wonder if there's any tests that capture the modern behavior that we previously disabled in Mono.

**Comment by kg** [comment]
cc @BrzVlad is this going to cause any problems for interp?

**Comment by steveisok** [comment]
This seems to have crashed on ios/tvos device runs. I ran @lambdageek's sample locally at least works for iossimulator. I'll try on a local device and see if I can get any info out of it.

File: src/mono/mono/metadata/class-init.c [hunk 1]
```diff
@@ -1039,18 +1039,14 @@ class_composite_fixup_cast_class (MonoClass *klass, gboolean for_ptr)
 	case MONO_TYPE_U2:
 		klass->cast_class = mono_defaults.int16_class;
 		break;
-	case MONO_TYPE_U4:
-#if TARGET_SIZEOF_VOID_P == 4
 	case MONO_TYPE_I:
```
Comment (jkotas) [inline src/mono/mono/metadata/class-init.c:1042]: `MONO_TYPE_BOOLEAN` -> `byte_class` normalization above looks suspect as well. I do not think regular CoreCLR does normalization like that anywhere.

File: src/mono/mono/metadata/class-init.c [hunk 2]
```diff
@@ -1039,18 +1039,14 @@ class_composite_fixup_cast_class (MonoClass *klass, gboolean for_ptr)
...
 	case MONO_TYPE_U:
-#endif
+		klass->cast_class = mono_defaults.int_class;
```
Comment (fanyang-mono) [inline src/mono/mono/metadata/class-init.c:1044]: Does `int_class` maps to 64bits when `TARGET_SIZEOF_VOID_P == 8`, and 32bits when `TARGET_SIZEOF_VOID_P == 4`?

Comment (lambdageek) [reply src/mono/mono/metadata/class-init.c:1044]: yes

**Comment by steveisok** [comment]
Weird, local device runs are fine.

File: src/mono/mono/metadata/class-init.c (hunk 1 above)
Comment (lambdageek) [reply src/mono/mono/metadata/class-init.c:1042]: @jkotas I think the `bool -> int8` fixup comes from when I misread I.8.7.2:

> A location type T is *compatible-with* a location type U if and only if one of the following holds.
> 1. T and U are not managed pointer types and T is *compatible-with* U according to the definition in §I.8.7.1.
> 2. T and U are both managed pointer types and T is *pointer-element-compatible-with* U.
> A managed pointer type T is *pointer-element-compatible-with* a managed pointer type U if and
> only if T has *verification type* V and U has *verification type* W and V is identical to W.

The *verification type* of a bool is defined in I.8.7 as

> 1. If the reduced type of T is:
>    a. int8 or bool, then its verification type is int8.


The problem is I think I applied the *verification type* rules to **un**managed pointer types (ie `bool*`) to managed pointer types (ie `bool&`)

**Review by fanyang-mono — APPROVED** [review]

**Comment by steveisok** [comment]
Appears there's some infra issues with the tvos and ios device legs. I validated system.runtime tests on a local device.

File: src/tests/Loader/classloader/Casting/Normalization.csproj [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
```
Comment (jkotas) [inline src/tests/Loader/classloader/Casting/Normalization.csproj:4]:
```suggestion
```

Comment (steveisok) [reply src/tests/Loader/classloader/Casting/Normalization.csproj]: That's what I get for copying ;-)

**Pull request merged**